---
title: "Fungi Sturgeon Eggs"
author: "Kristi Gdanetz MacCready"
date: '2022-05-02'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Documents/")
```

# Load libraries
```{r}
library(vegan)
library(tidyverse)
library(phyloseq)
library(broom)
library(agricolae)
library(cowplot)
theme_set(theme_cowplot())
set.seed(333)
source("~/Documents/GitHub/DenefLab-MicrobeMiseq/R/miseqR.R")
```

# Custom fuction
import indivual files, merge to create phyloseq object 
```{r}
import_mrg_physq = function(otu.table, 
                            map.file, 
                            tax.table, 
                            new.physq.obj){
  f.otu <- read.delim(otu.table, 
                     row.names=1, 
                     comment.char="!") %>% 
    as.matrix()  %>% 
    otu_table(taxa_are_rows = TRUE)
  
  f.mapping <- read.delim(map.file, 
                        row.names=1, 
                        comment.char="!") %>%
    as.data.frame() %>%
    sample_data()
  
  f.taxonomy <- read.delim(tax.table, 
                         header=TRUE, 
                         row.names=1) %>%
    as.matrix() %>%
    tax_table()
  
  new.physq.obj <- phyloseq(f.otu, f.mapping, f.taxonomy)
  return(new.physq.obj)
}
```

# Colors palette
```{r}
my_colors3 <- c('#8ed5f0','#ffb2b2',
                '#bc93d1','#ffe564',
                '#d2fe4c','#ff7fbf',
                '#7f0000','#004c66',
                '#995100','#3c1451',
                '#660033', '#2c4c00',
                '#ff7f7f','#33b5e5',
                '#ffbb33','#aa66cc',
                '#ff3298', '#99cc00',
                '#cc0000','#007299',
                '#cc6c00','#6b238e',
                '#b20058', '#436500',
                '#ff4444','#0099cc',
                '#ff8800','#9933cc',
                '#669900','#e50072')

# Color codes for figures: Tube A = #ff7f7f, Tube B = #33b5e5, 
# Tube C = #ffbb33, Tube D = #aa66cc, Tube F = #99cc00
```

# Import and filter data
```{r}
# make phyloseq object
physq.fungi <- import_mrg_physq("~/Documents/its.otu_tab.txt",
                            "~/Documents/fungi_sample_mapping.txt", 
                            "~/Documents/fungi_otus_taxonomy_05_sintax.txt", 
                            physq.fungi) #import 3 files need to create phyloseq object
physq.fungi = subset_taxa(physq.fungi, Domain == "Fungi") #remove OTUs that are not fungi
physq.fungi = subset_samples(physq.fungi, Short_name != "FUS") #remove postive control
ntaxa(physq.fungi) #272
nsamples(physq.fungi) #16

no.zero = prune_taxa(taxa_sums(physq.fungi) > 5, physq.fungi) #at least 5 reads per OTU
no.zero = prune_samples(sample_sums(no.zero) > 20, no.zero) #at least 20 reads per sample
ntaxa(no.zero) #218
nsamples(no.zero) #16

# pull out taxonomy table for Krona figure
fungi.tax.clean <- tax_table(no.zero)
# write.table(fungi.tax.clean, file='~/Documents/fungi_tax_clean.txt',
#             quote=FALSE, sep='\t')
```
Sample metadata
```{r}
meta_fung <- read_delim(file = "~/Documents/fungi_sample_mapping.txt", delim = "\t")
```


# Summary statistics
```{r}
sample_sum_df <- data.frame(sum = sample_sums(no.zero))

# Histogram of sample read counts
ggplot(sample_sum_df, aes(x = sum)) + 
  geom_histogram(color = "black", fill = "indianred", binwidth = 2500) +
  ggtitle("Distribution of sample sequencing depth") + 
  xlab("Read counts") +
  theme(axis.title.y = element_blank())
# summary statistics
smin <- min(sample_sums(no.zero)) #39
smean <- mean(sample_sums(no.zero)) #9038
smax <- max(sample_sums(no.zero)) #32798 
```

# Barplot code  
```{r}
# normalization
nsample = nsamples(no.zero)
sample_prop = transform_sample_counts(no.zero, function(x) x/sum(x))
taxa_sums(sample_prop)/nsample # this gives you the abundances
sum(taxa_sums(sample_prop)/nsample) # this should be 1 for each sample
sample_sums(sample_prop) # this should be 1 too

# format for neat plotting
taxa <- no.zero
glom <- tax_glom(taxa, 'Genus') # merge taxa
C <- prune_taxa(names(sort(taxa_sums(glom), decreasing=T)), taxa)

# transform to abundance 
norm <- transform_sample_counts(C, function(x) x/sum(x)) 
colSums(otu_table(norm)[,1:5])
norm.merge <- tax_glom(norm, "Genus")

# transform to relative abundace 
norm.merge = transform_sample_counts(norm.merge, function(x) (100*x/sum(x))/6)
colSums(otu_table(norm.merge)[,1:5])
```

Plot bar - do not use
```{r}
# generate barplot
plot_bar(norm.merge, fill='Genus', x='Tube', 
         title="Fungi") +
  ylab("Relative abundance (% of total sequences)") +
  scale_y_continuous(breaks=seq(0, 100, 25)) +
  geom_bar(aes(color=Genus, fill=Genus), stat='identity', position='stack') +
  scale_color_manual(values = my_colors3) + 
  scale_fill_manual(values = my_colors3)  +
  theme(axis.text = element_text(size = 10))
```

### Plot it 
```{r}
physq.fungi.f <- no.zero %>%
  tax_glom(taxrank = "Order") %>%                     # agglomerate at phylum level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt() %>%                                         # Melt to long format
  #filter(Abundance > 0.02) %>%                         # Filter out low abundance taxa
  arrange(Phylum)                                      # Sort data frame alphabetically by phylum

# Plot 
barplot_fungi <- physq.fungi.f %>%
  mutate(Tube = factor(Tube, ordered = TRUE,
         levels = c("A", "B", "C", "D"),
         labels = c("EBL", "LBL", "LBR", "FBR")) ) %>%
  ggplot(aes(x = Tube, y = Abundance, fill = Order)) + 
  geom_col() +
  scale_fill_manual(values = my_colors3) +
  scale_y_continuous(name = "Relative Abundance (% of total sequences)",
                      breaks = c(0, 1.5, 3, 4.5, 6), 
                     labels = c(0, 25, 50, 75, 100),
                     expand = c(0,0)) + 
  ylab("Relative Abundance (% of total sequences)") + xlab("Sample") +
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 10),
        legend.text = element_text(size = 8),  
        legend.title = element_text(size = 10)) 
barplot_fungi
#saveRDS(barplot_fungi, file = "~/Documents/Figures/barplot_fungi.RDS")
```

# Alpha diversity 
Calculate indices, format data frames for statistics and plotting 
```{r}
# remove empty taxa
no.zero <- prune_taxa(taxa_sums(no.zero) > 0, no.zero)

# Calculate Diversity Indices
alpha.fungi <- estimate_richness(no.zero, split = TRUE, 
                                measures = c("Observed", "Shannon"))
# # write to table
# write.table(alpha.fungi, file = "~/Documents/fungi_shannon.txt", sep = ",")

# add sample metadata to diversity indices
alpha.fungi2 <- alpha.fungi %>%
  rownames_to_column(var = "Sample_ID") %>%
  left_join(meta_fung, by = "Sample_ID") %>%
  # format for plotting 
  pivot_longer(cols = c("Observed", "Shannon"),
               values_to = "Measure",
               names_to = "Index") %>% 
  # relabel samples 
  mutate(Tube = factor(Tube, ordered = TRUE,
         levels = c("A", "B", "C", "D"),
         labels = c("EBL", "LBL", "LBR", "FBR")) ) %>%
  arrange(Tube, Index)

# summary statistics for plots
alpha.fungi.sum <- alpha.fungi2 %>%
  # relabel samples 
  mutate(Tube = factor(Tube, ordered = TRUE,
         levels = c("A", "B", "C", "D"),
         labels = c("EBL", "LBL", "LBR", "FBR")) ) %>%
  arrange(Tube, Index) %>% group_by(Tube, Index) %>%
  summarise(Mean_index = mean(Measure, na.rm = TRUE),
            StDev_index = sd(Measure, na.rm = TRUE)) 
```

### Tests 
test if diversity indices are different
```{r}
### Shannon index
ANOVA1 <- alpha.fungi2 %>%
  filter(Index == "Shannon") %>%
  aov(Measure~Tube, data = .)
summary(ANOVA1) #p=0.00457

# conduct Tukey's HSD test with agricolae and get letters for figure 
letters1 <- as.matrix(HSD.test(ANOVA1,"Tube", group=TRUE, alpha = 0.05)$groups) %>%
  as.data.frame() %>% #formatting the table with the letters
  rownames_to_column(var="Tube") %>%
  # relabel samples 
  mutate(Tube = factor(Tube, ordered = TRUE,
         levels = c("A", "B", "C", "D"),
         labels = c("EBL", "LBL", "LBR", "FBR")) ) %>% 
  mutate(Measure = round(as.numeric(as.character(Measure)), digits=2),
         groups = as.character(groups)) %>%
  rename(Shannon = Measure)

### observed ASVs index
ANOVA2 <- alpha.fungi2 %>%
  filter(Index == "Observed") %>%
  aov(Measure~Tube, data = .)
summary(ANOVA2) #p=0.328
letters2 <- as.matrix(HSD.test(ANOVA2,"Tube", group=TRUE, alpha = 0.05)$groups) %>%
  as.data.frame() %>% #formatting the table with the letters
  rownames_to_column(var="Tube") %>%
  # relabel samples 
  mutate(Tube = factor(Tube, ordered = TRUE,
         levels = c("A", "B", "C", "D"),
         labels = c("EBL", "LBL", "LBR", "FBR")) ) %>% 
  mutate(Measure = round(as.numeric(as.character(Measure)), digits=0),
         groups = as.character(groups)) %>%
  rename(Observed = Measure)

# # aggregate significant results
# sig.tukey <- tukey1 %>% 
#   filter(adj.p.value<0.5) %>% 
#   arrange(adj.p.value)
# sig.tukey
# # write.table(sig.tukey, file="~/Documents/fungi_shannon_tukey_results.txt", 
# #             sep="\t")
```

### Plot
```{r}
# generate plot
richplot_fungi <- alpha.fungi2 %>%
  # relabel samples 
  mutate(Tube = factor(Tube, ordered = TRUE,
                       levels = c("A", "B", "C", "D"),
                       labels = c("EBL", "LBL", "LBR", "FBR")) ) %>%
  ggplot(aes(x = Tube, y = Measure, color = Tube, Index)) +
  facet_wrap(.~Index, scales = "free") +
  # plot means and error bars
  geom_errorbar(data = alpha.fungi.sum, inherit.aes = FALSE, width = 0.5, color = "black",
                aes(x = Tube, ymin = Mean_index - StDev_index,
                    ymax = Mean_index + StDev_index)) +
  geom_crossbar(data = alpha.fungi.sum, inherit.aes = FALSE, fatten = 2, 
                aes(x = Tube, color = Tube,
                    y = Mean_index, ymin = Mean_index, ymax = Mean_index )) +
  # plot individual data points 
  geom_jitter(width = 0.25) +
  # set colors 
  scale_color_manual(name = "Sample", values = c("#ff7f7f", "#33b5e5", "#ffbb33", "#aa66cc", "#99cc00")) +
  # format axes
  ylab("Alpha Diversity Measure") + xlab("Sample") +
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 10))
richplot_fungi
```

# Beta diversity 
```{r}
# ordinate
dat_scale <- no.zero %>%
  scale_reads(round = "round") #scale to even depth
dat_nmds <- ordinate(
  physeq = dat_scale, 
  method = "NMDS", k=3, maxit=500, try=100,
  distance = "bray")
dat_nmds #stress=0.08047435 
```

### Plot
```{r}
# create dotplot 
nmds_fungi <- plot_ordination(
  physeq = dat_scale,
  ordination = dat_nmds,
  color = "Tube") +
  scale_color_manual("Sample",
                     values = c('#ff7f7f', '#33b5e5', '#ffbb33', '#aa66cc'), #'#99cc00'
                     labels = c("EBL", "LBL", "LBR", "FBR") ) +
  geom_point(aes(color = Tube), size = 4) +
  # reduce text size
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 10),
        legend.text = element_text(size = 8),  
        legend.title = element_text(size = 10) )
nmds_fungi
```

### Statistics 
```{r}
# Permanova and Betadisp 
dat_bray <- phyloseq::distance(dat_scale, method="bray") #calculate bray curtis distance matrix
# make a data frame from the sample_data
sampledf <- data.frame(sample_data(no.zero))

# Adonis test 
adon1 <- adonis(dat_bray~Tube, data=sampledf)
adon1 #p=0.002

# Homogeneity of dispersion test 
beta <- betadisper(dat_bray, sampledf$Tube)
permutest(beta) #p=0.008

# Anosim test 
ano_group <- get_variable(dat_scale, "Tube")
ano <- anosim(dat_bray, ano_group)
ano$signif #p=0.004
ano$statistic #R=0.4600227
```

# Export Figures
Diversity figure
```{r}
# get legend
leg3 <- nmds_fungi

figure3 <- plot_grid(richplot_fungi + theme(legend.position = "none"), 
                     nmds_fungi + theme(legend.position = "none"),
                     nrow = 1, ncol = 2, 
                     labels = c("A", "B"))
figure3
ggsave2(filename = "~/Documents/fish_eggs/Figures/figure3.svg", figure3,
        device = "svg", dpi = 300, width = 178, height = 100, units = "mm")
ggsave2(filename = "~/Documents/fish_eggs/Figures/legend3.svg", leg3,
        device = "svg", dpi = 300, width = 85, height = 85, units = "mm")
```

Bar plots
```{r}
#barplot_fungi <- readRDS(file = "~/Documents/barplot_fungi.RDS")
#barplot_oomy <- readRDS(file = "~/Documents/barplot_oomy.RDS")
leg1a <- get_legend(barplot_oomy )
leg1b <- get_legend(barplot_fungi)

figure1 <- plot_grid(barplot_oomy + theme(legend.position = "none"), 
                     leg1a,
                     barplot_fungi + theme(legend.position = "none"), 
                     leg1b, 
                     nrow = 1, ncol = 4, rel_widths = c(1, 0.5, 1, 0.5), 
                     labels = c("A", "", "B", ""))
ggsave2(filename = "~/Documents/fish_eggs/Figures/figure1.svg", figure1,
        device = "svg", dpi = 300, width = 178, height = 130, units = "mm")
```

Supplemental table
```{r}
# Tax table
fungi.tax.clean <- tax_table(no.zero) %>%
  as.data.frame() %>%
  rownames_to_column(var = "OTU")
# Shared table
fungi.otu.clean <- otu_table(no.zero) %>%
  as.data.frame() %>%
  rownames_to_column(var = "OTU")
# Combine
supp_tab2 <- inner_join(fungi.tax.clean, fungi.otu.clean, by = "OTU") %>%
  rename("EBL_1-3" = "KSFN210A13", "EBL_1-5" = "KSFN210A15", 
         "LBL_1-3" = "KSFN210B13", "LBL_1-5" = "KSFN210B15", "LBL_2-1" = "KSFN210B21", 
         "LBL_2-5" = "KSFN210B25", "LBL_3-3" = "KSFN210B33", "LBL_3-4" = "KSFN210B34", 
         "LBR_1-5" = "KSFN210C15", "LBR_2-5" = "KSFN210C25", "LBR_3-3" = "KSFN210C33",
         "FBR_3-3" = "KSFN108D33","FBR_1-3" = "KSFN210D13", "FBR_1-4" = "KSFN210D14", 
         "FBR_2-3" = "KSFN210D23", "FBR_2-4" = "KSFN210D24")
write_delim(supp_tab2, file = "~/Documents/Supplemental_Table2.txt",
            delim = "\t", col_names = TRUE, na = "NA")
```

-----
end 
